File: day/one/typescript/src/CalorieCounting.ts
import { readFile } from 'node:fs'
import { argv, exit } from 'node:process';
import { createPriorityQueue } from '@algorithm.ts/priority-queue';
import { assert } from 'node:console';

if (process.argv.length != 3) {
    console.error("usage: ts-node-esm " + __filename + " <input file>");
    exit(1);
}

const filepath = argv[2];

readFile(filepath, (err, data: Buffer) => {
    if (err) {
        throw err;
    }
    const contents = data.toString();

    const q = createPriorityQueue((x: number, y: number) => x - y);

    let total: number = 0;
    for (const line of contents.split("\n")) {
        if (line === "" || line === "\r") {
            q.enqueue(total);
            total = 0;
        } else {
            total += parseInt(line);
        }
    }

    let max = q.dequeue();
    let secondmax = q.dequeue();
    let thirdmax = q.dequeue();
    if (!(max && secondmax && thirdmax)) {
        assert(false, `max = ${max}, secondmax = ${secondmax} , thridmax = ${thirdmax}`);
        exit(1);
    }

    console.log(`The elf with the most calories is carrying ${max} calories`);
    console.log(`The top three elves are carrying a total of ${max + secondmax + thirdmax} calories `);
});

File: day/one/java/CalorieCounting.java
package day.one.java;

import java.io.File;
import java.io.FileNotFoundException;
import java.util.Comparator;
import java.util.PriorityQueue;
import java.util.Scanner;

@SuppressWarnings("java:S106") // command line utility
public class CalorieCounting {

    public static void main(String[] args) throws FileNotFoundException {
        if (args.length != 1) {
            System.err.println("Usage: java CalorieCounting <filename>");
            System.exit(1);
        }
        File f = new File(args[0]);
        try (Scanner sc = new Scanner(f)) {
            PriorityQueue<Integer> calories = new PriorityQueue<>(Comparator.reverseOrder());
            int total = 0;
            while (sc.hasNextLine()) {
                String line = sc.nextLine();
                if ("".equals(line)) {
                    calories.add(total);
                    total = 0;
                } else {
                    total += Integer.parseInt(line);
                }
            }

            int max = calories.poll();
            System.out.printf("The elf with the most calories is carrying %d calories%n", max);
            max += calories.poll();
            max += calories.poll();
            System.out.printf("The top three elves are carrying a total of %d calories %n", max);
        } catch (NumberFormatException e) {
            throw new IllegalArgumentException(e);
        }
    }
}